### PR TITLE
Apply input caching first before other scheduling operations

### PR DESCRIPTION
**Pull request by naoyam**
This makes sure all tensors including caches get the same scheduling operations applied.

See #663

### CODE REVIEW

**Comment by shmsong** [comment]
Fixes #663

**Comment by rdspring1** [comment]
I would just place ln. 1151 - 1165 @ ln.1214. We only use cache_after for persistent kernel on the fastest dimension.

  ```
    if (kHasOuterAxis) {
      // Fusion input castOp replaces cache_after
      // Determine if there are any casts or broadcast on fusion
      // inputs
      const auto& in_tv = ir_utils::filterByType<TensorView>(fusion->inputs());
      for (const auto input : in_tv) {
        if (input->getRootDomain().size() > 1) {
          // If pseudo-cache, skip cache after
          bool hasBroadcast = checkBroadcast(input) != nullptr;
          bool hasCast = checkCastOp(input) != nullptr;
          if (!hasBroadcast && !hasCast) {
            other_tv.push_back(input->cache_after());
          }
        }
      }
    }
```

**Comment by naoyam** [comment]
I changed it without understanding the context much, so please update it if necessary. One thing to note is that it needs to be done before calling `organizeAxes` and the cache tensors need to be inserted to `all_tv`.

**Comment by rdspring1** [comment]
I think the cache functions should have the option to replay transformations without applying computeAt.

`TransformReplay::replayCasP(consumer, producer, -1);`

**Comment by naoyam** [comment]
I actually tried that as well, and it wasn't enough in another fusion. I remember it looked like:

```
tv1[X, rY, ceilDiv(Z, Y)] computeAt (1) = reduce(tv0[X, Z])
```

In this case, `cache_after` does set computeAt for the cache tensor but does not do the split on the second axis since the position is 1. And the split of the second axis was important for the following parallelization. Here, `replayCasP` won't be helpful either. `replayPasC` with `tv1` should be done instead.

I agree that it would be great if cache_after/cache_before automatically could do the right transformation for the new cache tensor. However,  it's not trivial to infer what the right structure a cache tensor should look like, and I don't have a good idea other than avoiding the issue by caching before computeAt.

**Comment by rdspring1** [comment]
I want to replay the transformation after organizeAxes but before computeAt.

The cache tensor should look like its input.

**Comment by shmsong** [comment]
> I want to replay the transformation after organizeAxes but before computeAt.
> 
> The cache tensor should look like its input.

@rdspring1 Would moving the `cache_after` part down just before `computeAt` be ok for now? 
i.e. here:
[https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/scheduler.cpp#L1220](https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/scheduler.cpp#L1220)

`computeAt` will do the replay.

Sure `cache_after` was only used in `persistent` && `fastest_dim`.

File: torch/csrc/jit/codegen/cuda/scheduler.cpp [hunk 1]
```diff
@@ -1156,6 +1182,8 @@ void scheduleNormalization(
   const auto& in_tv = ir_utils::filterByType<TensorView>(fusion->inputs());
   const auto& out_tv = ir_utils::filterByType<TensorView>(fusion->outputs());
 
+  cacheInputs(fusion, rparams, reduction_tv, other_tv);
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: We want the cache afters only in below branch.

```cpp
if (rparams.fastest_dim) {
    const bool kHasOuterAxis = reduction_tv.front()->nDims() > 1;
    if (rparams.persistent_kernel) {
```

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/scheduler.cpp]: Yes, that condition is checked inside `cacheInputs`.

**Review by csarofeen — APPROVED** [review]

**Pull request merged**